### PR TITLE
pull request for HW3 continuing developing hiperglm

### PR DESCRIPTION
**Pull request by Mavis-Liang**

### CODE REVIEW

**Comment by Mavis-Liang** [comment]
Notes for Andrew:
I added test for Hessian (and the test passed); but there are still bugs in newton's method -- it converges at the first few iterations and the outputs aren't the same as BFGS (failing tests).

File: R/hiper_glm.R [hunk 1]
```diff
@@ -26,19 +29,6 @@ hiper_glm <- function(design, outcome, model='linear', option = list()){
...
-vcov.hglm <- function(hglm_out){
-  return(diag(length(hglm_out$coef)))
-}
-
-#' @export print.hglm
-print.hglm <- function(hglm_out){
-  return("printing...")
-}
-
+coef(hiper_glm(design, outcome, model = "logit", option = list(mle_solver="BFGS")))
```
Comment (chinandrew) [inline R/hiper_glm.R:32]: -1 

This doesn't need to be here and will lead to errors when importing.

Comment (chinandrew) [inline R/hiper_glm.R]: -1 

This doesn't need to be here and will lead to errors when importing.

File: R/hiper_glm.R [hunk 2]
```diff
@@ -2,21 +2,24 @@
...
-  supported_models <- c('linear')
-  if(!(model %in%  supported_models)){stop(sprintf("The model %s is not supported.",
-                                                   model))
+  supported_models <- c("linear", "logit")
+  if(!(model %in%  supported_models)){
+    stop(sprintf("The model %s is not supported.", model))
   }
 
 
   if (is.null(option$mle_solver)) {
```
Comment (chinandrew) [inline R/hiper_glm.R]: may have mentioned this in hw1, but you may want some logic here to error if an invalid solver is provided, since right now if someone specifies "gradient descent", itll do newton raphson and they won't know.

File: R/logistics_helper.R [hunk 3]
```diff
@@ -0,0 +1,72 @@
+## log-likelihood
+logit_loglik <- function(beta, X, Y){
+  p <- 1 / (1 + exp(-X %*% beta))
+  l <- -(t(Y) %*% log(p) + t(1 - Y) %*% log(1 - p))
+  return(l)
+}
+
+
+## gradient
```
Comment (chinandrew) [inline R/logistics_helper.R]: -1

These comments don't describe any more than the function names, if you don't want to write full function documentation I'd just remove them.

File: R/logistics_helper.R [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+logit_gradient <- function(beta, X, Y){
+  y_est <- 1 / (1 + exp(-X %*% beta))
+  gr = as.vector(t(X) %*% (y_est - Y))
+  return(gr)
+}
+
+## BFGS
+BFGS_finder_logit <- function(design, outcome, method){
```
Comment (chinandrew) [inline R/logistics_helper.R:39]: -2

This function is almost identical to your linear regression version. Better to write a single `bfgs()` function that takes in design, outcome, function, and gradient. Then you can pass in the desired likelikhood and gradient.

File: R/logistics_helper.R [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+## Hessian matrix
+hessian <- function(beta, X, Y){
+  P <- 1 / (1 + exp(-X %*% beta))
+  return(t(X) %*% diag(as.vector(P * (1 - P))) %*% X)
+}
+
+
+## Check convergence
+converged <- function(curr, prev, df){
+  tol <- qchisq(0.95, df, lower.tail = F) / 2
```
Comment (chinandrew) [inline R/logistics_helper.R]: -3

This is too large, for this assignment you can keep it simple and reuse the are_all_close() function with the default tolerances to check for convergence.

File: R/logistics_helper.R [hunk 6]
```diff
@@ -0,0 +1,72 @@
+## log-likelihood
+logit_loglik <- function(beta, X, Y){
+  p <- 1 / (1 + exp(-X %*% beta))
```
Comment (chinandrew) [inline R/logistics_helper.R]: -1

this is reused elsewhere, pull it out into an `expit()` function

File: R/logistics_helper.R [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+                       obj_fn, obj_grad,
+                       method = method,
+                       control = list(fnscale = -1))
+  return(BFGS$par)
+}
+
+
+## Hessian matrix
+hessian <- function(beta, X, Y){
+  P <- 1 / (1 + exp(-X %*% beta))
```
Comment (chinandrew) [inline R/logistics_helper.R]: -1

Be consistent with naming and try not to use capital letters. This is `p` elsewhere. Similarly for function name, everything else is `logit_loglik/grad`, but here you just have `hessian`

File: tests/testthat/test_grad.R [hunk 8]
```diff
@@ -15,19 +15,87 @@ test_that("linear model's analytical gradient is close to numerical one", {
...
+      x_m <- x
+      x_m[i] <- x_m[i] - dx
+      x_p <- x
+      x_p[i] <- x_p[i] + dx
+      numerical_grad[i,] <- (func(x_p) - func(x_m)) / (2 * dx)
+  }
+  return(numerical_grad)
+}
+
+test_that("Hessian matrix is alright", {
```
Comment (chinandrew) [inline tests/testthat/test_grad.R:73]: Be specific in your test names, what does 'alright' mean?

File: tests/testthat/test_grad.R [hunk 9]
```diff
@@ -15,19 +15,87 @@ test_that("linear model's analytical gradient is close to numerical one", {
   n_obs <- 32; n_pred <- 4
```
Comment (chinandrew) [inline tests/testthat/test_grad.R:15]: You'll probably want to rename this file since it's testing more than the gradient now.

File: tests/testthat/test_grad.R [hunk 10]
```diff
@@ -15,19 +15,87 @@ test_that("linear model's analytical gradient is close to numerical one", {
...
+    analytical_grad <- logit_gradient(regcoef, data$design, data$outcome)
+    numerical_grad <- approx_grad(loglik_func, regcoef)
+    grads_are_close <- are_all_close(
+      analytical_grad, numerical_grad, abs_tol = Inf, rel_tol = 1e-3
+    )
+  }
+  expect_true(grads_are_close)
+})
+
+approx_second_grad <- function(func, x, dx = .Machine$double.eps^(1/3)) {
```
Comment (chinandrew) [inline tests/testthat/test_grad.R:60]: put this in a helper function. Also, this is just checking the first derivative and not the second.

File: tests/testthat/test-mle-finder.R [hunk 11]
```diff
@@ -11,3 +11,17 @@ test_that("linalg and optim least-sq coincide", {
     coef(via_linalg_out), coef(via_bfgs_out), abs_tol = 1e-2, rel_tol = 1e-2
   ))
 })
+
+
+test_that("newton and bfgs outputs coincide on logit model", {
```
Comment (chinandrew) [inline tests/testthat/test-mle-finder.R:16]: -4

Test is failing. It's because of the above comments relating to implementing the negative log likelihood, and setting your threshold too loose for convergence.

File: R/logistics_helper.R [hunk 12]
```diff
@@ -0,0 +1,72 @@
+## log-likelihood
+logit_loglik <- function(beta, X, Y){
+  p <- 1 / (1 + exp(-X %*% beta))
+  l <- -(t(Y) %*% log(p) + t(1 - Y) %*% log(1 - p))
```
Comment (chinandrew) [inline R/logistics_helper.R]: You don't really need to use `l` here, you can just return this line

File: R/logistics_helper.R [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+  p <- ncol(X)
+  coefs <- rep(0, p)
+  loglik_curr <- logit_loglik(coefs, X, Y)
+  for (i in 1:maxIt) {
+    loglik_prev <- loglik_curr
+
+    ## Update the coefficients
+    gr <- as.matrix(logit_gradient(coefs, X, Y))
+    h <- hessian(coefs, X, Y)
+    coefs <- as.vector(coefs - solve(h) %*% gr)
```
Comment (chinandrew) [inline R/logistics_helper.R]: -2

Don't invert the matrix directly, use solve(A,b)

File: R/logistics_helper.R (hunk 12 above)
Comment (chinandrew) [inline R/logistics_helper.R]: You implemented the negative log likelihood (and respective gradients/hessians), so your BFGS is trying to maximize the wrong thing.

**Review by chinandrew — CHANGES_REQUESTED** [review]

**Comment by chinandrew** [comment]
> Notes for Andrew: I added test for Hessian (and the test passed); but there are still bugs in newton's method -- it converges at the first few iterations and the outputs aren't the same as BFGS (failing tests).

Thanks for the note, I've graded accordingly and let you know where your issue is.

File: R/logistics_helper.R [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+converged <- function(curr, prev, df){
+  tol <- qchisq(0.95, df, lower.tail = F) / 2
+  abs_diff <- abs(curr - prev)
+  within_atol <- abs_diff < tol
+  within_rtol <- (abs_diff < tol * max(abs(curr), abs(prev)))
+  return(within_atol && within_rtol)
+}
+
+## Newton's method main function
+newton <- function(X, Y, maxIt = 100){
```
Comment (chinandrew) [inline R/logistics_helper.R]: Late comment, but wanted to point out that `maxIt` doesn't  follow the same naming convention as the other variables.

File: R/logistics_helper.R (hunk 6 above)
Comment (chinandrew) [reply R/logistics_helper.R]: i saw you added expit(), but it was never used in the code

File: R/logistics_helper.R (hunk 3 above)
Comment (chinandrew) [reply R/logistics_helper.R]: there are two other isntances of this in the file that are missed

**Comment by chinandrew** [comment]
Final grade 90.5
